File: ios/Sources/FindUsEngine/SosEngine.swift
import Foundation

/// SOS gradient engine (port of core/sos.py): a device only holds the hops it
/// adopted itself; dedup is bounded.
public final class SosEngine {
    public let defaultTtlHops: Int
    public let defaultLifespan: Double?
    public let dedupWindow: Int

    public private(set) var events: [String: Sos] = [:]
    public private(set) var gradient: [String: [String: HopState]] = [:]
    public private(set) var adoptions = 0

    private var seen: Set<String> = []

    public init(defaultTtlHops: Int = 8, defaultLifespan: Double? = 900.0,
                dedupWindow: Int = 20000) {
        self.defaultTtlHops = defaultTtlHops
        self.defaultLifespan = defaultLifespan
        self.dedupWindow = dedupWindow
    }

    @discardableResult
    public func createSos(_ sosId: Any, origin: Any, t: Double, ttlHops: Int? = nil,
                          lifespan: Double? = nil) -> Sos {
        let sid = SosId(String(describing: sosId))
        let o = NodeId(String(describing: origin))
        let ttl = ttlHops ?? defaultTtlHops
        let life = lifespan ?? defaultLifespan
        let sos = Sos(id: sid, origin: o, createdAt: t, version: 1, ttlHops: ttl,
                      lifespanSeconds: life, expiresAt: life != nil ? t + life : nil)
        events[sid.value] = sos
        gradient[sid.value] = [o.value: HopState(hop: 0, version: 1, updatedAt: t)]
        return sos
    }

    public func isActive(_ sosId: Any, t: Double) -> Bool {
        guard let sos = events[String(describing: sosId)], sos.status == .active else {
            return false
        }
        guard let exp = sos.expiresAt else { return true }
        return t <= exp
    }

    public func activeSosIds(t: Double) -> [SosId] {
        events.values.filter {
            guard $0.status == .active else { return false }
            guard let exp = $0.expiresAt else { return true }
            return t <= exp
        }.map { $0.id }
    }

    @discardableResult
    public func expire(_ sosId: Any, t: Double) -> Bool {
        guard let sos = events[Self.key(for: sosId)] else { return false }
        sos.status = .expired
        gradient.removeValue(forKey: Self.key(for: sosId))
        return true
    }

    @discardableResult
    public func expireStale(t: Double) -> [SosId] {
        let dead = events.values.filter {
            guard $0.status == .active, let exp = $0.expiresAt else { return false }
            return t > exp
        }.map { $0.id }
        for sid in dead {
            events[sid.value]?.status = .expired
            gradient.removeValue(forKey: sid.value)
        }
        return dead
    }

    public func renew(_ sosId: Any, t: Double) -> Sos? {
        guard let sos = events[String(describing: sosId)], sos.status == .active else {
            return nil
        }
        sos.version += 1
        if sos.expiresAt != nil, let life = sos.lifespanSeconds {
            sos.expiresAt = t + life
        }
        return sos
    }

    /// Port of hear(): accepts only strictly-better routes.
    public func hear(_ sosId: Any, fromNode: Any, advertiserHop: Int, version: Int = 1,
                     t: Double, checkDuplicate: Bool = true) -> (Bool, Int?) {
        let sid = SosId(String(describing: sosId))
        let node = NodeId(String(describing: fromNode))
        guard isActive(sid, t) else { return (false, nil) }
        if checkDuplicate && markSeen(sid, version, String(describing: fromNode), advertiserHop) {
            return (false, nil)
        }
        guard let sos = events[sid.value] else { return (false, nil) }
        let candidate = advertiserHop + 1
        guard candidate <= sos.ttlHops else { return (false, nil) }
        if gradient[sid.value] == nil { gradient[sid.value] = [:] }
        let cur = gradient[sid.value]?[node.value]
        if let cur, version < cur.version { return (false, nil) }
        if let cur, candidate >= cur.hop { return (false, nil) }
        gradient[sid.value]?[node.value] = HopState(hop: candidate, version: version, updatedAt: t)
        adoptions += 1
        return (true, candidate)
    }

    private func markSeen(_ sid: SosId, _ version: Int, _ sender: String, _ hop: Int) -> Bool {
        let key = "\(sid.value)\u{0}\(version)\u{0}\(sender)\u{0}\(hop)"
        if seen.contains(key) { return true }
        if seen.count >= dedupWindow { seen.removeAll() }
        seen.insert(key)
        return false
    }

    public func hopOf(_ sosId: Any, node: Any) -> Int? {
        gradient[Self.key(for: sosId)]?[String(describing: node)]?.hop
    }

    static func key(for sosId: Any) -> String {
        if let s = sosId as? SosId { return s.value }
        return String(describing: sosId)
    }
}

File: android/app/src/main/kotlin/com/findus/crowdcompass/engine/SosEngine.kt
package com.findus.crowdcompass.engine

/**
 * SOS gradient engine port of core/sos.py. A device only ever holds the hops it
 * personally adopted (spec §53); dedup is bounded.
 */
class SosEngine(
    val defaultTtlHops: Int = 8,
    val defaultLifespan: Double? = 900.0,
    val dedupWindow: Int = 20000,
) {
    val events = linkedMapOf<String, Sos>()
    val gradient = linkedMapOf<String, LinkedHashMap<String, HopState>>()
    private val seen = linkedSetOf<String>()
    var adoptions = 0

    fun createSos(sosId: Any, origin: Any, t: Double, ttlHops: Int? = null,
                  lifespan: Double? = null): Sos {
        val sid = SosId(sosId.toString())
        val o = NodeId(origin.toString())
        val ttl = ttlHops ?: defaultTtlHops
        val life = lifespan ?: defaultLifespan
        val sos = Sos(sid, o, createdAt = t, version = 1, ttlHops = ttl,
            lifespanSeconds = life, expiresAt = if (life != null) t + life else null)
        events[sid.value] = sos
        gradient[sid.value] = linkedMapOf(o.value to HopState(0, 1, t))
        return sos
    }

    fun isActive(sosId: Any, t: Double): Boolean {
        val sos = events[sosId.toString()] ?: return false
        if (sos.status != SosStatus.ACTIVE) return false
        val exp = sos.expiresAt ?: return true
        return t <= exp
    }

    fun activeSosIds(t: Double): List<SosId> =
        events.values.filter {
            if (it.status != SosStatus.ACTIVE) return@filter false
            val exp = it.expiresAt
            exp == null || t <= exp
        }.map { it.id }

    fun expire(sosId: Any, t: Double): Boolean {
        val sos = events[sosId.toString()] ?: return false
        sos.status = SosStatus.EXPIRED
        gradient.remove(sosId.toString())
        return true
    }

    fun expireStale(t: Double): List<SosId> {
        val dead = events.values.filter {
            if (it.status != SosStatus.ACTIVE) return@filter false
            val exp = it.expiresAt
            exp != null && t > exp
        }.map { it.id }
        for (sid in dead) {
            events[sid.value]?.status = SosStatus.EXPIRED
            gradient.remove(sid.value)
        }
        return dead
    }

    fun renew(sosId: Any, t: Double): Sos? {
        val sos = events[sosId.toString()] ?: return null
        if (sos.status != SosStatus.ACTIVE) return null
        sos.version += 1
        if (sos.expiresAt != null && sos.lifespanSeconds != null) {
            sos.expiresAt = t + sos.lifespanSeconds
        }
        return sos
    }

    /** Port of hear(): accepts only strictly-better routes. */
    fun hear(sosId: Any, fromNode: Any, advertiserHop: Int, version: Int = 1,
             t: Double, checkDuplicate: Boolean = true): Pair<Boolean, Int?> {
        val sid = SosId(sosId.toString())
        val node = NodeId(fromNode.toString())
        if (!isActive(sid, t)) return false to null
        if (checkDuplicate && markSeen(sid, version, fromNode.toString(), advertiserHop)) {
            return false to null
        }
        val sos = events[sid.value]!!  // isActive implies present
        val candidate = advertiserHop + 1
        if (candidate > sos.ttlHops) return false to null
        val g = gradient.getOrPut(sid.value) { linkedMapOf() }
        val cur = g[node.value]
        if (cur != null && version < cur.version) return false to null
        if (cur != null && candidate >= cur.hop) return false to null
        g[node.value] = HopState(candidate, version, t)
        adoptions += 1
        return true to candidate
    }

    private fun markSeen(sid: SosId, version: Int, sender: String, hop: Int): Boolean {
        val key = "${sid.value}\u0000$version\u0000$sender\u0000$hop"
        if (key in seen) return true
        if (seen.size >= dedupWindow) seen.clear()  // bounded memory (reference semantics)
        seen.add(key)
        return false
    }

    fun hopOf(sosId: Any, node: Any): Int? =
        gradient[sosId.toString()]?.get(node.toString())?.hop
}